libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TemplatesRegisterExtensions.cs

```diff
@@ -0,0 +1,216 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using AdaptiveExpressions;
+using AdaptiveExpressions.Memory;
+using Microsoft.Bot.Builder.Dialogs.Adaptive.Generators;
+using Microsoft.Bot.Builder.LanguageGeneration;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive
+{
+    /// <summary>
+    /// Register lg templates into Expression.
+    /// Using prefix lg. to avoid the conflict with pre-built functions.
+    /// </summary>
+    public static class TemplatesRegisterExtensions
+    {
+        /// <summary>
+        /// Register Template functions.
+        /// </summary>
+        /// <param name="dialogContext">Dialog context.</param>
+        /// <param name="generator">Language generator.</param>
+        public static void RegisterTemplateFunctions(this DialogContext dialogContext, LanguageGenerator generator)
+        {
+            var lgm = dialogContext.Services.Get<LanguageGeneratorManager>();
+            var templateNames = GetTemplates(generator, dialogContext).Distinct();
+
+            var prebuildFunctionNames = Expression.Functions.Values.Select(u => u.Type);
+            foreach (var templateName in templateNames)
+            {
+                var functionType = prebuildFunctionNames.Contains(templateName) ? "lg." + templateName : templateName;
+                Expression.Functions.Add(functionType, new ExpressionEvaluator(
+                    functionType,
+                    (expression, state, options) =>
+                    {
+                        var currentLocale = GetCurrentLocale(state, options);
+
+                        if (state.TryGetValue(AdaptiveDialog.GeneratorIdKey, out var resourceId))
```

Comment:
>f (state.TryGetValue(AdaptiveDialog.GeneratorIdKey, out var resourceId)) [](http://example.com/codeflow?start=25&length=72)

resource.Id simply comes from name.  Why not?

Reply:
because it's "dynamic" and have *access control*, all functions are actually has the same evaluator. This evaluator will pick up generator based on current dialog's generator, and search name in this generator to make sure only the refered lg's templates is accessible.